manager/controllers/app/blueprint_controller.go

```diff
@@ -51,41 +52,87 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 
-	if blueprint.DeletionTimestamp == nil {
-		if !equality.Semantic.DeepEqual(&blueprint.Status, observedStatus) {
-			if err := r.Client.Status().Update(ctx, &blueprint); err != nil {
-				return ctrl.Result{}, errors.WrapWithDetails(err, "failed to update blueprint status", "status", blueprint.Status)
-			}
+	if !equality.Semantic.DeepEqual(&blueprint.Status, observedStatus) {
+		if err := r.Client.Status().Update(ctx, &blueprint); err != nil {
+			return ctrl.Result{}, errors.WrapWithDetails(err, "failed to update blueprint status", "status", blueprint.Status)
 		}
 	}
 
 	log.Info("blueprint reconcile cycle completed", "result", result)
 	return result, nil
 }
 
-func (r *BlueprintReconciler) deleteChartResource(ctx context.Context, kubeNamespace string, releaseName string) (ctrl.Result, error) {
-	rel, err := r.Helmer.Status(kubeNamespace, releaseName)
-	if err == nil && rel != nil {
-		_, _ = r.Helmer.Uninstall(kubeNamespace, releaseName)
+// reconcileFinalizers reconciles finalizers for Blueprint
+func (r *BlueprintReconciler) reconcileFinalizers(blueprint *app.Blueprint) error {
+	// finalizer
+	finalizerName := r.Name + ".finalizer"
+	hasFinalizer := utils.ContainsFinalizer(blueprint.GetFinalizers(), finalizerName)
+
+	// If the object has a scheduled deletion time, delete it and its associated resources
+	if !blueprint.DeletionTimestamp.IsZero() {
+		// The object is being deleted
+		if hasFinalizer { // Finalizer was created when the object was created
+			// the finalizer is present - delete the allocated resources
+			if err := r.deleteExternalResources(blueprint); err != nil {
+				return err
+			}
+
+			// remove the finalizer from the list and update it, because it needs to be deleted together with the object
+			ctrlutil.RemoveFinalizer(blueprint, finalizerName)
+
+			if err := r.Update(context.Background(), blueprint); err != nil {
+				return err
+			}
+		}
+		return nil
 	}
+	// Make sure this CRD instance has a finalizer
+	if !hasFinalizer {
+		ctrlutil.AddFinalizer(blueprint, finalizerName)
+		if err := r.Update(context.Background(), blueprint); err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-	return ctrl.Result{}, nil
+func getReleaseName(step app.FlowStep) string {
+	// we add the "r" character at the beginning of the release name, since it must begin with an alphabetic character
+	return "r" + utils.Hash(step.Name, 20)
 }
 
-func (r *BlueprintReconciler) applyChartResource(ctx context.Context, log logr.Logger, ref string, vals map[string]interface{}, kubeNamespace string, releaseName string) (ctrl.Result, error) {
+func (r *BlueprintReconciler) deleteExternalResources(blueprint *app.Blueprint) error {
+	for _, step := range blueprint.Spec.Flow.Steps {
+		releaseName := getReleaseName(step)
+		rel, err := r.Helmer.Status(blueprint.Namespace, releaseName)
+		if err == nil && rel != nil {
+			_, _ = r.Helmer.Uninstall(blueprint.Namespace, releaseName)
+		}
+	}
+	return nil
```

Comment:
This seems like a best effort to delete, ignoring errors. Either:
1. Declare that this is a best effort and then you can avoid the call to Status (counting on the removal of the namespace to do the trick in most cases)
2. Track errors and if you encountered one then on the end of the loop don't return nil but rather an error (better IMHO)

Reply:
Done.